big_scape/comparison/binning.py

```diff
@@ -129,6 +129,50 @@ def add_records(self, record_list: list[BGCRecord]):
         if None in self.record_ids:
             raise ValueError("Region in bin has no db id!")
 
+    def cull_singletons(self, cutoff: float):
+        """Culls singletons for given cutoff, i.e. records which have either no edges
+        in the database, or all edges have a distance above/equal to the cutoff"""
+
+        if not DB.metadata:
+            raise RuntimeError("DB.metadata is None")
+
+        distance_table = DB.metadata.tables["distance"]
+
+        # get all distances in the table below the cutoff
+        select_statement = (
+            select(distance_table.c.region_a_id, distance_table.c.region_b_id)
+            .where(
+                distance_table.c.region_a_id.in_(self.record_ids)
+                | distance_table.c.region_b_id.in_(self.record_ids)
+            )
+            .where(distance_table.c.distance < cutoff)
+            .where(distance_table.c.weights == self.weights)
+        )
+
+        edges = set(DB.execute(select_statement).fetchall())
+
+        # get all record_ids in the edges
+        filtered_record_ids = set()
+        for edge in edges:
+            filtered_record_ids.update(edge)
+
+        records_to_remove = []
+        record_ids_to_remove = set()
+
+        # TODO: can we clean up this code?
```

Comment:
You can just replace the record ids with the set you just generated, I think.
Then you can use that to generate a new list of source records. e.g.

```python
self.record_ids = filtered_record_ids
self.source_records = [record for record in self.source_records if record._db_id in filtered_record_ids]
```